lib/node_modules/@stdlib/random/base/mt19937/lib/factory.js

```diff
@@ -0,0 +1,246 @@
+*      products derived from this software without specific prior written
+*      permission.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+* A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+* CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+* EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+* PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+* PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+* LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+* NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+* ```
+*/
+
+'use strict';
+
+// MODULES //
+
+var max = require('@stdlib/math/base/special/max');
+var isCollection = require('@stdlib/assert/is-collection');
+var setReadOnly = require('@stdlib/utils/define-read-only-property');
+var isPositiveInteger = require('@stdlib/assert/is-positive-integer').isPrimitive;
+var UINT32_MAX = require('@stdlib/constants/math/uint32-max');
+var FLOAT64_MAX_SAFE_INTEGER = require('@stdlib/constants/math/float64-max-safe-integer');
+var randuint32 = require('./rand_uint32.js');
+
+
+// VARIABLES //
+
+var NORMALIZATION_CONSTANT = UINT32_MAX >>> 0; // asm type annotation
+var NORMALIZED_NORMALIZATION_CONSTANT = 1.0 / (FLOAT64_MAX_SAFE_INTEGER + 1.0);
+var MAX_SEED = UINT32_MAX >>> 0; // asm type annotation
+var N = 624;
+var M = 397;
+var UPPER_MASK = 0x80000000; /* Most significant w-r bits */
+var LOWER_MASK = 0x7fffffff; /* Least significant r bits */
+var MATRIX_A = 0x9908b0df;
+var mag01 = [ 0x0, MATRIX_A ];
+
+
+// MAIN //
+
+/**
+* Returns a 32-bit Mersenne Twister pseudorandom number generator.
+*
+* @param {PositiveInteger} [seed] - pseudorandom number generator seed
+* @throws {TypeError} must provide a positive integer
+* @throws {RangeError} must provide a positive integer less than the maximum signed 32-bit integer
+* @returns {Function} Mersenne Twister PRNG
+*
+* @example
+* var mt19937 = factory();
+*
+* var v = mt19937();
+* // returns <number>
+*
+* @example
+* // Return a seeded Mersenne Twister PRNG:
+* var mt19937 = factory( 1234 );
+*
+* var v = mt19937();
+* // returns 822569775
+*/
+function factory( seed ) {
+	var seedIsArray;
+	var state;
+	var mti;
+	var mt;
+	var s;
+	var i;
+	var j;
+	var k;
+
+	if ( arguments.length ) {
+		if ( isCollection( seed ) ) {
+			seedIsArray = true;
+			i = 1;
+			j = 0;
+			k = max( N, seed.length );
+			state = 19650218 >>> 0; // asm type annotation
+		} else {
+			if ( !isPositiveInteger( seed ) ) {
+				throw new TypeError( 'invalid argument. Must provide a positive integer. Value: `' + seed + '`.' );
+			}
+			if ( seed >= MAX_SEED ) {
```

Comment:
I think I asked this before, but what is the reason that `seed` cannot equal `UINT32_MAX`?

Reply:
It can be equal. I changed the value of `MAX_SEED` constant defined above.

Reply:
I realize that the constant `MAX_SEED` need not be defined and can be simply replaced by `UINT32_MAX`. But I reckon it's more readable and obvious with the extra variable.

Reply:
Understood. Am okay with that.